csrc/quantization/fp8/common.cu

```diff
@@ -118,12 +115,62 @@ __global__ void scaled_fp8_quant_kernel(c10::Float8_e4m3fn* __restrict__ out,
   }
 
   // Handle the remaining elements if num_elems is not divisible by 4
-  for (int i = num_vec_elems * 4 + tid; i < num_elems;
-       i += blockDim.x * gridDim.x) {
+  for (int i = num_vec_elems * 4 + tid; i < num_elems; i += step) {
     out[i] = scaled_fp8_conversion(input[i], inverted_scale);
   }
 }
 
+template <typename scalar_t>
+__global__ void scaled_fp8_quant_kernel(c10::Float8_e4m3fn* __restrict__ out,
+                                        const scalar_t* __restrict__ input,
+                                        const float* __restrict__ scale,
+                                        int64_t num_elems) {
+  int tid = blockDim.x * blockIdx.x + threadIdx.x;
+
+  // Invert the scale so that we can use multiplications to avoid expensive
+  // division.
+  const float inverted_scale = 1.0f / (*scale);
+
+  scaled_fp8_conversion_vec(out, input, inverted_scale, num_elems, tid,
+                            blockDim.x * gridDim.x);
+}
+
+template <typename scalar_t>
+__global__ void dynamic_per_token_scaled_fp8_quant_kernel(
+    c10::Float8_e4m3fn* __restrict__ out, float* __restrict__ scale,
+    scalar_t const* __restrict__ input, const int hidden_size) {
+  int const tid = threadIdx.x;
+  int const token_idx = blockIdx.x;
+  float absmax_val = 0.0f;
+
+  for (int i = tid; i < hidden_size; i += blockDim.x) {
+    float const x = static_cast<float>(input[token_idx * hidden_size + i]);
+    absmax_val = max(absmax_val, fabs(x));
+  }
+
+  float const block_absmax_val_maybe = blockReduceMax(absmax_val);
+  __shared__ float block_absmax_val;
+  if (tid == 0) {
+    block_absmax_val = block_absmax_val_maybe;
+    scale[token_idx] = block_absmax_val / FP8_E4M3_MAX;
+  }
+  __syncthreads();
+
+  float const inverted_scale = FP8_E4M3_MAX / block_absmax_val;
+  bool const vectorize_conversions = hidden_size % 4 == 0;
```

Comment:
Why only vectorize conversions if the hidden_size is divisible by 4? `scaled_fp8_conversion_vec` has a loop to "Handle the remaining elements if num_elems is not divisible by 4"

Reply:
The input and output pointers need to be aligned  8-byte and 4-byte addresses respectively.  Each row of the input is processed by a single block. 
The FP8 output data-type requires that hidden-size (each token/row of the output has hidden-size elements) be a multiple of 4, otherwise the 2nd, 3rd and 4th output rows will no the 4-byte aligned. The argument is similar for float16, bfloat16, float32 input datatypes.

Reply:
@comaniac

Reply:
Why didn't we run into this with the original `scaled_fp8_quant_kernel`?

Reply:
the `scaled_fp8_quant_kernel` has a single scale that is applied to all of the input values and it doesn't have to process row-by-row. For quant purposes, that kernel can treat the inputs and output as one big row of values (the "first row" case from above).